File: riveredge-frontend/src/apps/kuaiplm/services/production-file.ts
/**
 * 生产文件 API（R-06）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/production-files';

export type ProductionFileCatalogKind = 'pe_production' | 'rd_tool';

export type ProductionFileStatus =
  | 'draft'
  | 'pending'
  | 'effective'
  | 'obsolete'
  | 'rejected';

export type ProductionFileType =
  | 'burn'
  | 'laser'
  | 'bluetooth_ir'
  | 'aoi'
  | 'label_template'
  | 'other_pe'
  | 'rd_burn_tool'
  | 'rd_prod_test'
  | 'other_rd';

export interface ProductionFile {
  uuid: string;
  id: number;
  file_code: string;
  catalog_kind: ProductionFileCatalogKind;
  file_type: ProductionFileType | string;
  title: string;
  process_code?: string | null;
  process_name?: string | null;
  product_model?: string | null;
  project_id?: number | null;
  project_code?: string | null;
  project_name?: string | null;
  release_date?: string | null;
  version: string;
  status: ProductionFileStatus;
  file_uuid?: string | null;
  file_name?: string | null;
  checksum?: string | null;
  change_summary?: string | null;
  remarks?: string | null;
  issued_by_name?: string | null;
  issued_at?: string | null;
  receiver_names?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  obsolete_at?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by?: number | null;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface ProductionFilePayload {
  catalog_kind: ProductionFileCatalogKind;
  file_type: string;
  title: string;
  version?: string;
  file_code?: string;
  process_code?: string | null;
  process_name?: string | null;
  product_model?: string | null;
  project_id?: number | null;
  project_code?: string | null;
  project_name?: string | null;
  release_date?: string | null;
  file_uuid?: string | null;
  file_name?: string | null;
  checksum?: string | null;
  change_summary?: string | null;
  remarks?: string | null;
}

export interface ProductionFileVersion {
  id: number;
  uuid: string;
  file_id: number;
  file_code: string;
  version: string;
  status: string;
  is_effective: boolean;
  is_production_effective: boolean;
  title: string;
  file_type?: string | null;
  release_date?: string | null;
  file_uuid?: string | null;
  file_name?: string | null;
  change_summary?: string | null;
  created_by_name?: string | null;
  created_at?: string;
}

export interface ProductionFileAccessLog {
  id: number;
  uuid: string;
  file_id: number;
  file_code: string;
  version_id?: number | null;
  version?: string | null;
  action: string;
  actor_name?: string | null;
  receiver_names?: string | null;
  remark?: string | null;
  created_at: string;
}

function unwrapList<T>(res: unknown): { items: T[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as T[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const productionFileApi = {
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    catalog_kind?: string;
    file_type?: string;
    process_code?: string;
    product_model?: string;
    project_id?: number;
    production_view?: boolean;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList<ProductionFile>(res);
  },
  get: async (id: number) => {
    const res = await api.get(`${BASE}/${id}`);
    return res as ProductionFile;
  },
  getDownloadUrl: async (
    id: number,
    options?: { version_id?: number; production_view?: boolean },
  ) => {
    const res = await api.get(`${BASE}/${id}/download`, {
      params: {
        version_id: options?.version_id,
        production_view: options?.production_view ? true : undefined,
      },
    });
    return res as { preview_url: string; file_name?: string | null; version_id?: number | null };
  },
  create: async (data: ProductionFilePayload) => {
    const res = await api.post(BASE, data);
    return res as ProductionFile;
  },
  update: async (id: number, data: Partial<ProductionFilePayload>) => {
    const res = await api.put(`${BASE}/${id}`, data);
    return res as ProductionFile;
  },
  submit: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/submit`);
    return res as ProductionFile;
  },
  approve: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/approve`);
    return res as ProductionFile;
  },
  reject: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/reject`);
    return res as ProductionFile;
  },
  revise: async (
    id: number,
    data: {
      version?: string;
      change_summary?: string;
      file_uuid?: string;
      file_name?: string;
      release_date?: string | null;
      title?: string;
    },
  ) => {
    const res = await api.post(`${BASE}/${id}/revise`, data);
    return res as ProductionFile;
  },
  obsolete: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/obsolete`);
    return res as ProductionFile;
  },
  issue: async (id: number, data: { receiver_names: string; remark?: string; version_id?: number }) => {
    const res = await api.post(`${BASE}/${id}/issue`, data);
    return res as ProductionFile;
  },
  recordAccess: async (
    id: number,
    data: { action: 'view' | 'download'; version_id?: number },
    productionView = false,
  ) => {
    const res = await api.post(`${BASE}/${id}/access`, data, {
      params: { production_view: productionView },
    });
    return res as ProductionFileAccessLog;
  },
  listVersions: async (id: number, productionView = false) => {
    const res = await api.get(`${BASE}/${id}/versions`, {
      params: { production_view: productionView },
    });
    const unwrapped = unwrapList<ProductionFileVersion>(res);
    const r = (res || {}) as Record<string, unknown>;
    return {
      ...unwrapped,
      audience: String(r.audience || ''),
      can_view_history: Boolean(r.can_view_history),
    };
  },
  listAccessLogs: async (id: number, params?: { action?: string; skip?: number; limit?: number }) => {
    const res = await api.get(`${BASE}/${id}/access-logs`, { params });
    return unwrapList<ProductionFileAccessLog>(res);
  },
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};

File: riveredge-backend/src/apps/kuaiplm/schemas/production_file.py
"""生产文件中心 Schema（R-06）。"""

from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ProductionFileCreate(BaseModel):
    catalog_kind: str = Field(..., description="pe_production / rd_tool")
    file_type: str = Field(..., min_length=1, max_length=40)
    title: str = Field(..., min_length=1, max_length=200)
    version: str = Field("A0", min_length=1, max_length=30)
    file_code: Optional[str] = Field(None, max_length=50)
    process_code: Optional[str] = Field(None, max_length=80)
    process_name: Optional[str] = Field(None, max_length=200)
    product_model: Optional[str] = Field(None, max_length=120)
    project_id: Optional[int] = Field(None, description="研发项目ID（与 project_code 二选一或皆空）")
    project_code: Optional[str] = Field(None, max_length=50, description="项目代号（存量项目手填）")
    project_name: Optional[str] = Field(None, max_length=200, description="项目名称（手填时可空，默认同代号）")
    release_date: Optional[date] = None
    file_uuid: Optional[str] = Field(None, max_length=36)
    file_name: Optional[str] = Field(None, max_length=200)
    checksum: Optional[str] = Field(None, max_length=128)
    change_summary: Optional[str] = None
    remarks: Optional[str] = None


class ProductionFileUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    file_type: Optional[str] = Field(None, min_length=1, max_length=40)
    process_code: Optional[str] = Field(None, max_length=80)
    process_name: Optional[str] = Field(None, max_length=200)
    product_model: Optional[str] = Field(None, max_length=120)
    release_date: Optional[date] = None
    file_uuid: Optional[str] = Field(None, max_length=36)
    file_name: Optional[str] = Field(None, max_length=200)
    checksum: Optional[str] = Field(None, max_length=128)
    change_summary: Optional[str] = None
    remarks: Optional[str] = None
    version: Optional[str] = Field(None, min_length=1, max_length=30)


class ProductionFileReviseRequest(BaseModel):
    version: Optional[str] = Field(None, max_length=30)
    change_summary: Optional[str] = None
    file_uuid: Optional[str] = Field(None, max_length=36)
    file_name: Optional[str] = Field(None, max_length=200)
    checksum: Optional[str] = Field(None, max_length=128)
    release_date: Optional[date] = None
    title: Optional[str] = Field(None, max_length=200)


class ProductionFileIssueRequest(BaseModel):
    receiver_names: str = Field(..., min_length=1, max_length=500)
    remark: Optional[str] = Field(None, max_length=500)
    version_id: Optional[int] = None


class ProductionFileAccessRequest(BaseModel):
    action: str = Field(..., description="view / download")
    version_id: Optional[int] = None
    remark: Optional[str] = Field(None, max_length=500)


class ProductionFileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    file_code: str
    catalog_kind: str
    file_type: str
    title: str
    process_code: Optional[str] = None
    process_name: Optional[str] = None
    product_model: Optional[str] = None
    project_id: Optional[int] = None
    project_code: Optional[str] = None
    project_name: Optional[str] = None
    release_date: Optional[date] = None
    version: str
    status: str
    file_uuid: Optional[str] = None
    file_name: Optional[str] = None
    checksum: Optional[str] = None
    change_summary: Optional[str] = None
    remarks: Optional[str] = None
    issued_by: Optional[int] = None
    issued_by_name: Optional[str] = None
    issued_at: Optional[datetime] = None
    receiver_names: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    obsolete_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None


class ProductionFileListResponse(BaseModel):
    items: List[ProductionFileResponse]
    total: int


class ProductionFileVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    file_id: int
    file_code: str
    version: str
    status: str
    is_effective: bool
    is_production_effective: bool
    title: str
    file_type: Optional[str] = None
    release_date: Optional[date] = None
    file_uuid: Optional[str] = None
    file_name: Optional[str] = None
    checksum: Optional[str] = None
    change_summary: Optional[str] = None
    effective_at: Optional[datetime] = None
    obsolete_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None


class ProductionFileVersionListResponse(BaseModel):
    items: List[ProductionFileVersionResponse]
    total: int
    audience: str
    can_view_history: bool


class ProductionFileAccessLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    file_id: int
    file_code: str
    version_id: Optional[int] = None
    version: Optional[str] = None
    action: str
    actor_user_id: Optional[int] = None
    actor_name: Optional[str] = None
    receiver_names: Optional[str] = None
    remark: Optional[str] = None
    created_at: datetime


class ProductionFileAccessLogListResponse(BaseModel):
    items: List[ProductionFileAccessLogResponse]
    total: int


class ProductionFileDownloadResponse(BaseModel):
    preview_url: str = Field(..., description="带 token 的下载预览 URL")
    file_name: Optional[str] = Field(None, description="文件名快照")
    version_id: Optional[int] = Field(None, description="版本行 ID")
